ui/src/store.js

```diff
@@ -0,0 +1,186 @@
+          inputPurse = purse;
+          if (inputPurse === outputPurse) {
+            outputPurse = null;
+          }
+        }
+        if (fieldNumber === 1) {
+          outputPurse = purse;
+          if (outputPurse === inputPurse) {
+            inputPurse = null;
+          }
+        }
+        return { ...state, inputPurse, outputPurse, freeVariable };
+      },
+    },
+    swapInputs(state) {
+      const { inputPurse, outputPurse, inputAmount, outputAmount } = state;
+      return {
+        ...state,
+        inputPurse: outputPurse,
+        outputPurse: inputPurse,
+        inputAmount: outputAmount,
+        outputAmount: inputAmount,
+      };
+    },
+    changeAmount: {
+      // Map positional arguments.
+      create: (amount, fieldNumber, freeVariable = null) => ({
+        amount,
+        fieldNumber,
+        freeVariable,
+      }),
+      reducer(state, { amount, fieldNumber, freeVariable }) {
+        let { inputAmount, outputAmount } = state;
+        if (fieldNumber === 0) {
+          inputAmount = amount;
+        }
+        if (fieldNumber === 1) {
+          outputAmount = amount;
+        }
+        return { ...state, inputAmount, outputAmount, freeVariable };
+      },
+    },
+    createOffer: {
+      // Map positional arguments.
+      create: (
+        instanceHandleBoardId,
+        installationHandleBoardId,
+        invitationDepositId,
+        inputAmount,
+        outputAmount,
+        inputPurse,
+        outputPurse,
+      ) => ({
+        instanceHandleBoardId,
+        installationHandleBoardId,
+        invitationDepositId,
+        inputAmount,
+        outputAmount,
+        inputPurse,
+        outputPurse,
+      }),
+      reducer(
+        state,
+        {
+          instanceHandleBoardId,
+          installationHandleBoardId,
+          invitationDepositId,
+          inputAmount,
+          outputAmount,
+          inputPurse,
+          outputPurse,
+        },
+      ) {
+        const offer = {
+          // JSONable ID for this offer.  Eventually this will be scoped to
+          // the current site.
+          id: Date.now(),
+
+          // TODO: get this from the invitation instead in the wallet. We
+          // don't want to trust the dapp on this.
+          instanceHandleBoardId,
+          installationHandleBoardId,
+
+          proposalTemplate: {
+            give: {
+              In: {
+                // The pursePetname identifies which purse we want to use
+                pursePetname: inputPurse.pursePetname,
+                value: inputAmount,
+              },
+            },
+            want: {
+              Out: {
+                pursePetname: outputPurse.pursePetname,
+                value: outputAmount,
+              },
+            },
+            exit: { onDemand: null },
+          },
+        };
+
+        // Create an invitation for the offer and on response, send the proposed
+        // offer to the wallet.
+        doFetch(
```

Comment:
Are we supposed to have external calls here? I thought this was intended to only manipulate state. I think I've gotten an answer about this before, but I've forgotten why we are doing this.

Reply:
> Are we supposed to have external calls here? I thought this was intended to only manipulate state.

It's a mishmash of state and effects, as is `contexts/Application.jsx`.

> I think I've gotten an answer about this before, but I've forgotten why we are doing this.

I believe the answer was that if we used full-blown Redux, we would use middleware to do the effects.  Since we rolled our own application context, we don't have middleware, so we have to plumb in effects directly.  Other options would be more indirect like somehow having a component adapt to a change in state to fire off a request.